net/net.sh

```diff
@@ -232,11 +232,11 @@ build() {
   echo "Build took $SECONDS seconds"
 }
 
-SOLANA_HOME="\$HOME/solana"
-CARGO_BIN="\$HOME/.cargo/bin"
-
 startCommon() {
   declare ipAddress=$1
+  remoteHome="$(ssh "${sshOptions[@]}" "$ipAddress" "echo \$HOME")"
+  SOLANA_HOME="${remoteHome}/solana"
+  CARGO_BIN="${remoteHome}/.cargo/bin"
```

Comment:
**minor**

There are a number of commands that rely on `~` working.
For example:

```bash
      mkdir -p $SOLANA_HOME/config;
      rm -rf ~/config;
      mv $SOLANA_HOME/config ~;
      rm -rf $SOLANA_HOME;
```

and

```bash
    rsync -vPrc -e "ssh ${sshOptions[*]}" "$SOLANA_ROOT"/solana-release/bin/* "$ipAddress:$CARGO_BIN/"
    rsync -vPrc -e "ssh ${sshOptions[*]}" "$SOLANA_ROOT"/solana-release/version.yml "$ipAddress:~/"
```

So, it might also be fine to use `~` everywhere:

```suggestion
  SOLANA_HOME="~/solana"
  CARGO_BIN="~/.cargo/bin"
```

P.S.  When quoted, bash does not do tile expansion on the assignment.
For example:
```bash
mkdir 1
a=~/1
b="~/1"
echo "[$a]"
# [/home/user/1]
echo "[$b]"
# [~/1]
```

Reply:
I think the issue is that we don't want $HOME to expand to the local environment's home directory, but expand it to the remote user's home directory. This part of the code hasn't changed for quite some time. So not sure what caused it to break.

Reply:
> yeah, the problem is that `rysnc` knows how to expand `~`, but not shell vars. all of the `ssh`'d remote commands get piped through `bash` on the other end, where vars are expanded. `rsync` just gets an arg string with a `$` that it doesn't care about

Not sure, I fully understand what you refer to as "the problem".
There is already an invocation of `rsync` that relies on correct expansion of `~`:
```bash
rsync -vPrc -e "ssh ${sshOptions[*]}" "$SOLANA_ROOT"/solana-release/version.yml "$ipAddress:~/"
```
And so if `$remoteCargoBin` is `~/.cargo/bin`, then the other command:
```bash
rsync -vPrc -e "ssh ${sshOptions[*]}" "$SOLANA_ROOT"/solana-release/bin/* "$ipAddress:$remoteCargoBin/"
```
will be seen by `rsync` as
```bash
rsync -vPrc -e "ssh sshOptionsValue" ..../net/../solana-release/bin/* "1.1.1.1:~/.cargo/bin/"
```

And bash code sent inside `ssh` will have variables substituted on the client side.  So the remote side will see
```bash
mkdir -p ~/solana/config;
```
rather than
```bash
mkdir -p $remoteSolanaHome/config;
```

But, it is not a big difference.
Just save one extra ssh round trip :)